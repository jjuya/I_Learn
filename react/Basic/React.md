## React

- http://webframeworks.kr/tutorials/react/getting-started/



### 페이스북은 왜 React를 만들었는가?

- [React 공식 문서](http://facebook.github.io/react/docs/why-react-ko-KR.html)
- (전) React 팀이었던 [Pete Hunt가 쓴 글](http://facebook.github.io/react/blog/2013/06/05/why-react.html)

- **지속해서 데이터가 변화**하는 대규모 애플리케이션을 구축하기 위해서
  - 유지보수를 위하여 모델 데이터를 변경하는 로직과 뷰 로직을 분리
  - '뷰 데이터 관리 도구'를 통해 Separation of Concern(관심사의 분리)를 달성
- React의 경우 public하게 사용할 수 있는 API는 몇 개 되지 않는다.
  - JSX를 통해 어떤 형태로 뷰 데이터가 보여져야 하는지 **선언적으로 기술**
  - React는 중간 과정이 아닌 결과물을 기술하는 것
- 관심사의 분리 + 선언적인 API



### React는 무엇인가

#### 뷰 레이어

- MVC, MV* 패던에서 뷰 레이어만을 담당

- 어떤 형태의 모델이 사용될 것인지에 대한 가정을 하기 않는다.

  - 아무 라이브러리를 사용해도 무방
  - 작은 앱이라면 굳이 라이브러리 사용할 필요가 없다.

  >데이터를 서버에서 가져온다거나 (ajax 요청), 데이터들을 조작한다거나 (underscore/lodash) location에 따라 다른 컴포넌트를 마운트해준다거나 (`Backbone.Router`, `angular-route`) 하는 기능은 직접 구현하거나, 상황에 맞는 라이브러리를 사용 

- 참고

  - [Flux 아키텍쳐](http://facebook.github.io/flux/)
  - [Redux](https://github.com/rackt/redux) 



#### 컴포넌트를 통한 뷰 작성

- Reactive한 단방향의 데이터 흐름
  - Reactive 하다 : 상태(일반적인 의미에서의 state)가 바뀌면 뷰도 함꼐 업데이트
  - 단방향 : 한 방향으로 데이터가 흐른다.
    - 상위 컴포넌트(Owner)에서 하위 컴포넌트(Ownee)로
    - 데이터는 React에서 **prop**이라고 지칭 
    - JSX에서는 HTML의 attribute처럼 작성
- 컴포넌트 라이프사이클 제공



##### 예제 : 쇼핑몰에 쇼핑카트

```react
// 참고: 동적으로 생성되는 모델의 경우 보통은 `Array.prototype.map` 을 이용하여
// 배열 형태 모델을 작성하고, key를 배정하지만 글의 목적에 맞게 간단하게 쓰기 위하여
// 정적으로 작성합니다.
var ShoppingCart = React.createClass({
    render () {
        return <div>
            <ShoppingItem name="kimchi" available={true} />
            <ShoppingItem name="rice" available={true} />
            <ShoppingItem name="curry" available={false} />
        </div>
    }
});

// 참고: 기존 html 엘리먼트가 아닌 한, 모든 커스텀 컴포넌트들의 이름은 대문자로
// 쓰여져야 하며 지켜지지 않으면 에러가 납니다.
var ShoppingItem = React.createClass({
    render () {
        return <div>
            <div>
                상품명: {this.props.name},
                구입가능: {this.props.available ? '가능' : '불가능'}
            </div>
        </div>
    }
});
```

- 컴포넌트는 `ShoppingCart -> ShoppingItem` 형태
- `<ShoppingItem name="kimchi" available={true} />`
  - **JSX**라는 치환 문법
  - HTML attribute처럼 써준 것이 하위 컴포넌트로 상태를 주입
- Owner에서 `React.CreateElement` 함수 호출로 새 컴포넌트 인스턴스 생성
- `React.render` : DOM 컨테이너 위에 Render되는 순간



#### Virtual DOM과 Reconciliation

- HTML Element들을 Virtual DOM을 이용해 표현 

  - Virtual DOM은 가상의 HTML Element들을 가지고 있다가, (재)렌더링을 하면 필요한 부분만 업데이트(DOM 조작)

- **Always re-render on update**

  - jQuery 등을 이용하여 애플리케이션을 작성하면 모델이 업데이트되었을 때 필요한 컴포넌트들만 셀렉터를 이용하여 업데이트를 해주는 코드를 작성 
  - React는 작성가자 원하는 결과물만을 선언적으로 작성

- Reconciliation (비교조정) 

  - 가지고 있는 Virtual DOM 트리를 비교하면서 필요한 부분만 업데이트

  - [React 공식 문서](http://facebook.github.io/react/docs/reconciliation-ko-KR.html)

  - [React의 diffing 알고리즘에 대한 글](http://calendar.perfplanet.com/2013/diff/)

    > 동기 DOM 배치 오퍼레이션을 매 업데이트마다 수행하는 것보다는 낫겠지만, 글을 보면 업데이트 전, 업데이트 후의 가상 엘리먼트 트리를 비교해야 하는 문제가 있습니다. 이 문제는 가장 최신의 알고리즘도 `O(n^3)`의 시간 복잡도를 가지고 있는 수준이므로, 여러가지 가정들을 통해 알고리즘의 복잡도를 `O(n)`까지 낮췄다고 설명 

- 뷰 업데이트 로직은 거의 신경쓰지 않고 모델 데이터 관리와 결과물만 기술 

- 가상화라는 것은 기본적으로 브라우저 구현에 코드가 의존하는 것을 넘어선다는 것을 의미

  - 브라우저의 구현 디테일 차이들을 덜 신경써도 된다는 장점



### React에 대한 오해

#### React는 빠르다

- 비교조정 알고리즘은 `O(n^3)`에서 `O(n)`까지 시간 복잡도를 끌어올리긴 했지만, 비교하는 것 외에도 많은 관리 작업을 수행
  - React는 기본적으로 빠르지 않다.
- 해주는 것에 비해서는 빠르다.
- 최적화를 잘 해주면 빨라진다.
  - 기본적으로 런타임에서 invariant 등의 개발을 도와주는 기능들이 존재
  - 이런 부분을 프로덕션 빌드를 통해 제거

#### JSX는 템플릿 언어다

- JSX는 EcmaScript로 치환되는 간단한 치환/확장 언어

  - [E4X에 영향](http://blog.vjeux.com/2013/javascript/jsx-e4x-the-good-parts.html)을 받아 만들어짐
  -  함수 호출 식으로 써도 됨

- React 컴포넌트와 React.DOM 가상 엘리먼트 생성자들은 트랜스파일러를 통해 `React.createElement`함수 콜으로 치환

  ```react
  <div foo={0} bar={'baz'} />
  
  // 치환 후
  React.createElement('div', { foo: 0, bar: 'baz' });
  ```

- [Handlebars](http://guides.emberjs.com/v1.10.0/templates/displaying-a-list-of-items/) 나 [Jinja2](http://jinja.pocoo.org/docs/dev/)와 같이 자체적인 if-else, 반복문, 조건적 표현 블럭 등의 제어구조를 가지고 있지 않다.

  - EcmaScript 표현식들을 `{}` 안에 써준다는 정도



https://www.slideshare.net/floydophone/react-preso-v2