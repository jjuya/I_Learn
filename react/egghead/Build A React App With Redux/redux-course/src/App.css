.App {
    width: 400px;
    height: 700px;

    margin: auto;

    border: 1px solid rgba(128, 128, 128, 0.53);

    text-align: center;
}

.App-logo {
    animation: App-logo-spin infinite 20s linear;
    height: 80px;
}

.App-header {
    background-color: #222;
    height: 150px;
    padding: 20px;
    color: white;
}

.App-title {
    font-size: 1.5em;
}

.App-intro {
    font-size: large;
}

@keyframes App-logo-spin {
    from { transform: rotate(0deg); }
    to { transform: rotate(360deg); }
}

.Todo-App {
    padding: 25px 50px 10px 50px;

    overflow-y: hidden;
}

.input-text {
    width: 300px;
    height: 35px;

    padding: 0px;

    border: 1px solid rgba(128, 128, 128, 0.53);
    border-radius: 3px;
}

.Todo-List {
    align-items: left;
}

ul {
    width: 300px;

    padding: 0px;

    text-align: left;
}

li {
    list-style: none;
}