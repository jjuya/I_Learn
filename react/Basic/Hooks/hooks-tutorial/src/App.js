import React, { Component } from 'react'
import logo from './logo.svg'
import './App.css'

import Example from './components/Example'

class App extends Component {
  render() {
    return <Example />
  }
}

export default App
