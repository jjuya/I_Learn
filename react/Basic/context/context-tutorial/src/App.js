import React from 'react'
import { Row, Col } from 'antd'

import LeftPane from './components/LeftPane'
import RightPane from './components/RightPane'
import { SampleProvider } from './contexts/sample'
import { AnotherProvider } from './contexts/another'
import Counter from './components/Counter'

const AppProvider = ({ contexts, children }) =>
  contexts.reduce(
    (prev, context) =>
      React.createElement(context, {
        children: prev
      }),
    children
  )

const App = () => {
  return (
    <AppProvider contexts={[SampleProvider, AnotherProvider]}>
      <Row type="flex" gutter={16} justify="space-around">
        <Col span={12}>
          <LeftPane />
        </Col>
        <Col>
          <RightPane span={12} />
        </Col>
      </Row>
      <Counter />
    </AppProvider>
  )
}

export default App
