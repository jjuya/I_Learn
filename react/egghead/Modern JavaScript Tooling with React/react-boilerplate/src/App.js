import React from 'react'

class App extends React.Component {
  state = {
    count: 0
  }

  onIncrease = () => {
    this.setState(state => ({ count: state.count + 1 }))
  }

  onDecrease = () => {
    this.setState(state => ({ count: state.count - 1 }))
  }

  render() {
    // debugger
    return (
      <div>
        <h1>Hello World!!</h1>
        <h2>count: {this.state.count}</h2>
        <button onClick={this.onIncrease}>+</button>
        <button onClick={this.onDecrease}>-</button>
      </div>
    )
  }
}

export default App
