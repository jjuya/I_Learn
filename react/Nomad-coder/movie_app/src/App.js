import React, { Component } from 'react';
import './App.css';
import Movie from "./components/Movie";

// Render: componentWillMount() -> render() -> componentDidMount()
// Update componentWillReceiveProps() -> shouldComponentUpdate() -> componentWillUpdate() -> render() -> componentDidUpdate()

class App extends Component {
    state = {}

    componentDidMount() {
        this._getMovies()
    }

    _callApi = () => {
        // response : fetch의 결과물
        // const result = fetch("URL")
        //     .then(response => console.log("response"))
        //     .catch(error => console.log("Error!!"))


        return fetch("https://yts.am/api/v2/list_movies.json?sort_by=rating")
            .then(response => response.json())
            .then(json => json.data.movies)
            .catch(error => console.log(error))
    }

    _getMovies = async () => {
        const movies = await this._callApi()

        this.setState({
            movies
        })
    }

    _renderMovies = () => {
        const movies = this.state.movies.map(movie => {
            return (
                <Movie
                    title={movie.title_english}
                    poster={movie.large_cover_image}
                    key={movie.id}
                    genres={movie.genres}
                    synopsis={movie.synopsis}
                />

        )})


        return movies
    }

    render() {
        const { movies } = this.state

        return (
            <div className={movies ? "APP" : "APP--loading"}>
                {movies ? this._renderMovies() : "Loading..."}
            </div>
        );
    }
}

export default App;
