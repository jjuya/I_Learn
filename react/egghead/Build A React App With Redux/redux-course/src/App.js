import React, { Component } from 'react';
import {connect} from 'react-redux'
import logo from './logo.svg';
import './App.css';

import {bindActionCreators} from 'redux';
import {updateCurrent} from "./reducers/todo";

import TodoForm from './components/TodoForm'
import TodoList from './components/TodoList'
class App extends Component {
  render() {
    return (
      <div className="App">
        <div className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Welcome to React with Redux</h1>
        </div>

          <div className="Todo-App">
              <TodoForm
                  currentTodo={this.props.currentTodo}
                  changeCurrent={this.props.updateCurrent}
              />

              <TodoList
                  todos={this.props.todos}
              />

          </div>
      </div>
    );
  }
}

// export default App;
const mapStateToProps = (state) => state
const mapDispatchToProps = (dispatch) => bindActionCreators({updateCurrent}, dispatch)
const ConnectApp = connect(mapStateToProps, mapDispatchToProps)(App)
export default ConnectApp