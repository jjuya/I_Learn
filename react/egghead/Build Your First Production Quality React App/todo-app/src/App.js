import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

import PropTypes from 'prop-types'

import { TodoForm, TodoList, Footer } from "./components/todo";
import { addTodo, findById, generateId, toggleTodo, updateTodo, removeTodo, filterTodos } from './lib/todoHelpers'
import { pipe, partial } from './lib/utils'
import { loadTodos, createTodo, saveTodo, destroyTodo } from './lib/todoService'

class App extends Component {

    state = {
        todos: [],
        currentTodo: ''
    }

    static contextTypes = {
        route: PropTypes.string
    }

    componentDidMount() {
        loadTodos()
            .then(todos => this.setState({todos}))
    }

    handleInputChange = (e) => {
        this.setState({
            currentTodo: e.target.value
        })
    }

    handleSubmit = (e) => {
        e.preventDefault()

        if (this.state.currentTodo) {
            const newId = generateId()
            const newTodo = {id: newId, name: this.state.currentTodo, isComplete: false}
            const updatedTodos = addTodo(this.state.todos, newTodo)

            this.setState({
                todos: updatedTodos,
                currentTodo: ''
            })

            createTodo(newTodo)
                .then(() => console.log('Todo added'))
        }
    }

    handleEmptySubmit = (e) => {
        e.preventDefault()

        this.setState({
            errorMessage: 'Please supply a todo name'
        })
    }

    handleToggle = (id) => {
        const getToggledTodo = pipe(findById, toggleTodo)
        const updated = getToggledTodo(id, this.state.todos)
        const getUpdatedTodos = partial(updateTodo, this.state.todos)
        const updatedTodos = getUpdatedTodos(updated)

        this.setState({todos: updatedTodos})

        saveTodo(updated)
            .then(() => this.showTempMessage('Todo Updated'))
    }

    handleRemove = (id, e) => {
        e.preventDefault()

        const updatedTodos = removeTodo(this.state.todos, id)

        this.setState({todos: updatedTodos})

        destroyTodo(id)
    }

    showTempMessage = (msg) => {
        this.setState({message: msg})
        setTimeout(() => this.setState({message: ''}), 2500)
    }

    render() {
        const submitHandler = this.state.currentTodo ? this.handleSubmit : this.handleEmptySubmit

        const displayTodos = filterTodos(this.state.todos, this.context.route)

        return (
            <div className="App">
                <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo" />
                    <h1 className="App-title">React Todos</h1>
                </header>

                <div className="Todo-App">
                    {this.state.errorMessage && <span className='error'>{this.state.errorMessage}</span>}
                    <TodoForm
                        handleInputChange={this.handleInputChange}
                        handleSubmit={submitHandler}
                        currentTodo={this.state.currentTodo}
                    />

                    <TodoList
                        handleToggle={this.handleToggle}
                        handleRemove={this.handleRemove}
                        todos={displayTodos}
                    />
                </div>

                <Footer />
            </div>
        );
    }
}

export default App;
