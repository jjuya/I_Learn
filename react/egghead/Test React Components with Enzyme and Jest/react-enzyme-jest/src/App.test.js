import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

// Shallow rendering
describe('<App />', () => {
    const wrapper = shallow(<App />)

    it('should render App', () => {
        console.log(wrapper.debug())
    })

    it('should contain 1 p', () => {
        expect(wrapper.find('p').length).toBe(1)
    })

    it('should contain class name is App-intro', () => {
        expect(wrapper.find('.App-intro').exists()).toBe(true)
    })

    it('should have 3 children in ul tag ', () => {
        expect(wrapper.find('ul').children().length).toBe(3)
    })

    it('should have ul class name is tyler ', () => {
        expect(wrapper.find('ul').hasClass('tyler')).toBe(true)
    })

})

