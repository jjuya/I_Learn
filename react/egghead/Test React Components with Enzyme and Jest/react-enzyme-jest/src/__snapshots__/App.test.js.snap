// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Mount rendering matches the snapshot 1`] = `
<App>
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h1
        className="App-title"
      >
        Welcome to React
      </h1>
    </header>
    <p
      className="App-intro"
    >
      Hello react!
    </p>
    <ul
      className="tyler"
    >
      <li>
        test1
      </li>
      <li>
        test2
      </li>
      <li>
        test3
      </li>
    </ul>
  </div>
</App>
`;

exports[`<App /> Shallow rendering matches the snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
  </header>
  <p
    className="App-intro"
  >
    Hello react!
  </p>
  <ul
    className="tyler"
  >
    <li>
      test1
    </li>
    <li>
      test2
    </li>
    <li>
      test3
    </li>
  </ul>
</div>
`;
